ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByOperator.java

```diff
@@ -561,17 +590,25 @@ private void flush(boolean all) throws HiveException {
             maxHashTblMemory/1024/1024,
             gcCanary.get() == null ? "dead" : "alive"));
       }
+      int avgAccess = computeAvgAccess();
 
       /* Iterate the global (keywrapper,aggregationbuffers) map and emit
        a row for each key */
       Iterator<Map.Entry<KeyWrapper, VectorAggregationBufferRow>> iter =
           mapKeysAggregationBuffers.entrySet().iterator();
       while(iter.hasNext()) {
         Map.Entry<KeyWrapper, VectorAggregationBufferRow> pair = iter.next();
+        if (!all && avgAccess >= 1) {
+          // Retain entries when access pattern is > than average access
+          if (pair.getValue().getAccessCount() > avgAccess) {
```

Comment:
Yes, for now there is no plan to make this pluggable. This is mainly for the rollup queries, where it works best with LRU.

Reply:
@ashutoshc this conversation was still not resolved - I was waiting for a response; I think we could have improved further on this patch just by changing it a little bit.

@rbalamohan  we are batch removing from the cache elements here; which does not happen in regular LRU stuff.

if we have {{K}} cache slots; and start the stream with an element which is there for say {{N*K}} times ; that will raise the bar to retain a new cache element during flush to {{N}}.

I think the counters of the retained entries should be reset to 0 at least - it will increase it's effectiveness and neutralize long-term memory effects - like the above

Reply:
For rollup, it should be the other way around. If retained entries are reset to zero, it could potentially get evicted in next iteration polluting the map. I guess the confusion is with strict LRU impl here. Intent is not to implement strict LRU, and evict one at time (earlier impl with LinkedHashMap followed that, but ends up with L1 cache misses and ends up with heavy object tracking). Current implementation is light weight, but does not change the pattern of evicting 10% on entries. Just adds logic to retain heavily accessed entries. 

Potential further optimization is to reuse the evicted entities and reuse them for pooling (e.g for every keywrapper gets cloned internally via copyKey() in the map. This causes high mem pressure on certain queries). This can be added as an additional optimization in follow up jira.

Reply:
I don't think we should get to L1 cache stuff here; I'm ok with not having an LRU, we might not need to do that - let's stay on logical level:

I'm saying that if in a round the hit count of an element is not reduced or reset to zero (when it's being kept for next round) those keys could retain their places for a long time because of very old cache hits - and they will keep their place in the cache; this could lead to a case in which all the new elements (n/2) added to the cache are evicted and why very old entries which are not even getting any hits will keep half of the cache filled.

Reply:
>> keys could retain their places for a long time because of very old cache hits - and they will keep their place in the cache

This depends on incoming data and would be the worst case scenario similar to earlier implementation.  However, there is a corner case (again depending on data) that large number of entries in the map exceeds the average threshold which could prevent 10% flushing limit. Adding the reset would help preventing this. I will create a follow up ticket on this.

Reply:
https://issues.apache.org/jira/browse/HIVE-23917